allennlp/tests/end_to_end/empty_list_test.py

```diff
@@ -0,0 +1,218 @@
+        embedded_body_tokens = self.embedder(body)
+        # size: <batch_size, 2 * emb_dim>
+        embedded_body = self.fixed_length_embedding(body_mask, embedded_body_tokens)
+
+        # Embed name (in isolation)
+
+        # size: <batch_size, sequence_len>
+        name_mask = get_text_field_mask(entity_name)
+        # size: <batch_size, sequence_len, emb_dim>
+        embedded_name_tokens = self.embedder(entity_name)
+        # size: <batch_size, 2 * emb_dim>
+        embedded_name = self.fixed_length_embedding(name_mask, embedded_name_tokens)
+
+        # Extract embedded spans from the body
+
+        # size: <batch_size, mentions_count, sequence_len>
+        mentions_mask = get_text_field_mask(entity_mentions, num_wrapping_dims=1)
+        # size: <batch_size, mentions_count, sequence_len, emb_dim>
+        embedded_mentions_tokens = self.embedder(entity_mentions)
+        # size: [<batch_size, 1, 2 * emb_dim>]
+        embedded_spans_list = []
+        for i in range(mentions_mask.size(1)):
+            embedded_spans_tmp = self.fixed_length_embedding(
+                    mentions_mask[:, i, :],
+                    embedded_mentions_tokens[:, i, :, :]
+            ).unsqueeze(1)
+            embedded_spans_list.append(embedded_spans_tmp)
+        # size: <batch_size, mentions_count, 2 * emb_dim>
+        embedded_spans = torch.cat(embedded_spans_list, 1)
+
+        # size: <batch_size>
+        name_match_score = torch.nn.functional.cosine_similarity(embedded_body, embedded_name)
+
+        # size: <batch_size, 2 * emb_dim, mentions_count>
+        transposed_embedded_spans = embedded_spans.transpose(1, 2)
+        # Note: Real model normalizes to give cosine similarity.
+        # size: <batch_size, mentions_count>
+        span_match_scores = torch.bmm(embedded_body.unsqueeze(1), transposed_embedded_spans).squeeze(1)
+        # size: <batch_size, mentions_count>
+        masked_span_match_scores = span_match_scores * (mentions_mask[:, :, 0] != 0).float()
+        # Aggregate with max to get single score
+        # size: <batch_size>
+        span_match_score = masked_span_match_scores.max(dim=-1)[0].squeeze(-1)
+
+        # Combine name match and span match scores.
+        return {'score': name_match_score + span_match_score}
+
+class EmptyListTest(AllenNlpTestCase):
+    def test_empty_list_can_be_tensorized(self):
+        token_indexers = {"tokens": SingleIdTokenIndexer()}
+        tokenizer = WordTokenizer()
+        tokens = tokenizer.tokenize("Foo")
+        text_field = TextField(tokens, token_indexers)
+        list_field = ListField([text_field.empty_field()])
+        fields = {'list': list_field, 'bar': TextField(tokenizer.tokenize("BAR"), token_indexers)}
+        instance = Instance(fields)
+        vocab = Vocabulary.from_instances([instance])
+        instance.index_fields(vocab)
+        instance.as_tensor_dict()
+
+    # A batch with entirely empty lists.
+    def test_end_to_end_broken_without_fix(self):
```

Comment:
This test name will be totally meaningless in a few months once we've forgotten about this PR.  Can you name it something more descriptive of what it's actually testing?  Same with the test below.  Also, I feel like the salience stuff was very helpful for getting me to understand what the motivation was for this, but I'm not sure it's necessary to have it in the test.  I believe you just largely copied that code from your separate repo, anyway.  What do you think of just removing it and having these two tests be similar to what you have in the test above, with a comment explaining in text an actual use case for this behavior (to make it easy, you could just copy my rephrasing)?